File: repos/koi-web/src/lib/repositories/cbt_session_repository.ts
import { db } from '../db';
import type { CBTSession } from '../../models/index';

export class CBTSessionRepository {
  async save(session: CBTSession): Promise<void> {
    await db.cbtSessions.put(session);
  }

  async getById(id: string): Promise<CBTSession | undefined> {
    return db.cbtSessions.get(id);
  }

  async getByUser(userId: string): Promise<CBTSession[]> {
    return db.cbtSessions.where('userId').equals(userId).sortBy('startedAt');
  }

  async getCompletedByUser(userId: string): Promise<CBTSession[]> {
    const all = await this.getByUser(userId);
    return all.filter((s) => s.completedAt !== undefined);
  }

  async getUnsynced(): Promise<CBTSession[]> {
    return db.cbtSessions.where('synced').equals(0).toArray();
  }

  async markSynced(id: string): Promise<void> {
    await db.cbtSessions.update(id, { synced: true });
  }

  async complete(
    id: string,
    opts: { qualityScore?: number; userInsights?: string } = {},
  ): Promise<void> {
    await db.cbtSessions.update(id, {
      completedAt: new Date(),
      ...(opts.qualityScore !== undefined && { quality_score: opts.qualityScore }),
      ...(opts.userInsights !== undefined && { user_insights: opts.userInsights }),
    });
  }

  async delete(id: string): Promise<void> {
    await db.cbtSessions.delete(id);
  }
}

export const cbtSessionRepository = new CBTSessionRepository();

File: repos/koi-backend/flutter_lib/lib/storage/repositories/cbt_session_repository.dart
import 'package:hive/hive.dart';
import '../hive_manager.dart';
import '../../models/hive_cbt_session_model.dart';

class CBTSessionRepository {
  Box<HiveCBTSession> get _box => HiveManager.cbtSessions;

  Future<void> saveSession(HiveCBTSession session) async =>
      _box.put(session.id, session);

  HiveCBTSession? getSession(String id) => _box.get(id);

  List<HiveCBTSession> getByUser(String userId) {
    return _box.values
        .where((s) => s.userId == userId)
        .toList()
      ..sort((a, b) => a.startedAt.compareTo(b.startedAt));
  }

  List<HiveCBTSession> getCompletedByUser(String userId) {
    return _box.values
        .where((s) => s.userId == userId && s.completedAt != null)
        .toList()
      ..sort((a, b) => a.startedAt.compareTo(b.startedAt));
  }

  List<HiveCBTSession> getUnsynced() =>
      _box.values.where((s) => !s.synced).toList();

  Future<void> markSynced(String id) async {
    final session = _box.get(id);
    if (session == null) return;
    session.synced = true;
    await session.save();
  }

  Future<void> completeSession(String id, {double? qualityScore, String? userInsights}) async {
    final session = _box.get(id);
    if (session == null) return;
    session.completedAt = DateTime.now();
    if (qualityScore != null) session.qualityScore = qualityScore;
    if (userInsights != null) session.userInsights = userInsights;
    await session.save();
  }

  Future<void> deleteSession(String id) async => _box.delete(id);
}
